### PR TITLE
Improve MIR match generation for ranges

### PR DESCRIPTION
**Pull request by sinkuu**
Improves MIR match generation to rule out ranges/values distinct from the range that has been tested. e.g., for this code:

```rust
match x {
    0..=5 if b => 0,
    6..=10 => 1,
    _ => 2,
}
```

MIR (before):

```rust
bb0: { ...; _4 = Le(const 0i32, _1); switchInt(move _4) -> [false: bb6, otherwise: bb5]; }
bb1: { _3 = const 0i32; goto -> bb8; }
bb2: { _6 = _2; switchInt(move _6) -> [false: bb6, otherwise: bb1]; } // If `!b`, jumps to test if `6 <= x <= 10`.
bb3: { _3 = const 1i32; goto -> bb8; }
bb4: { _3 = const 2i32; goto -> bb8; }
bb5: { _5 = Le(_1, const 5i32); switchInt(move _5) -> [false: bb6, otherwise: bb2]; }
bb6: { _7 = Le(const 6i32, _1); switchInt(move _7) -> [false: bb4, otherwise: bb7]; }
bb7: { _8 = Le(_1, const 10i32); switchInt(move _8) -> [false: bb4, otherwise: bb3]; }
```

MIR (after):
```rust
bb0: { ...; _4 = Le(const 0i32, _1); switchInt(move _4) -> [false: bb5, otherwise: bb6]; }
bb1: { _3 = const 0i32; goto -> bb8; }
bb2: { _6 = _2; switchInt(move _6) -> [false: bb4, otherwise: bb1]; } // If `!b`, jumps to `_ =>` arm.
bb3: { _3 = const 1i32; goto -> bb8; }
bb4: { _3 = const 2i32; goto -> bb8; }
bb5: { _7 = Le(const 6i32, _1); switchInt(move _7) -> [false: bb4, otherwise: bb7]; }
bb6: { _5 = Le(_1, const 5i32); switchInt(move _5) -> [false: bb5, otherwise: bb2]; }
bb7: { _8 = Le(_1, const 10i32); switchInt(move _8) -> [false: bb4, otherwise: bb3]; }
```

cc #29623

### CODE REVIEW

**Comment by sinkuu** [comment]
Hmm, I need more local testing.

**Comment by sinkuu** [comment]
Imaginary part of a false edge always points to `Candidate.next_candidate_pre_binding_block`, which is populated in `Builder::match_expr`, so it won't be affected.

Essential<sup>*</sup> diff of `SimplifyBranch-initial.before` for that example:

```diff
@@ -58,7 +58,7 @@
         switchInt(move _6) -> [false: bb10, otherwise: bb1];
     }
     bb10: {
-        falseEdges -> [real: bb11, imaginary: bb5];
+        falseEdges -> [real: bb6, imaginary: bb5];
     }
     bb11: {
         _7 = Le(const 6i32, _1);
```

(\* this PR changes the order of BBs, so I swapped a few of them to reduce amount of diff)

**Review by oli-obk — COMMENTED** [review]
I think it would be a good idea to add a mir-opt test ensuring that we don't regress this in the future, but these tests are such a pain to keep updated that we should punt this to the future.

This is probably not going to affect the perf tests as it's unlikely that we have any doing such range matches in a perf critical path. Have you done any benchmarks locally? Just curious, does not affect this PR at all.

File: src/librustc_mir/build/matches/test.rs [hunk 1]
```diff
@@ -607,8 +638,70 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
...
+                if (test_lo, test_hi, test_end) == (pat_lo, pat_hi, pat_end) {
+                    resulting_candidates[0]
+                        .push(self.candidate_without_match_pair(
+                            match_pair_index,
+                            candidate,
+                        ));
+                    return true;
+                }
+
+                let no_overlap = (|| {
```
[the remark this answers is not in the document]
Comment (sinkuu) [reply src/librustc_mir/build/matches/test.rs]: `try` syntax seems to require not only `#![feature(try_block)]` but also `--edition 2018` for `try` keyword.

**Comment by sinkuu** [comment]
> Have you done any benchmarks locally?

No.

**Comment by oli-obk** [comment]
@rust-timer build f8b8ae37e663a3f881c18a6eae916a8d26570327

**Pull request merged**